sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/DecorrelateInnerQuerySuite.scala

```diff
@@ -283,6 +283,51 @@ class DecorrelateInnerQuerySuite extends PlanTest {
     check(innerPlan, outerPlan, correctAnswer, Seq(y <=> y, x === a, y === z))
   }
 
+  test("union in correlation path") {
+    val outerPlan = testRelation2
+    val innerPlan =
+      Union(
+        Filter(And(OuterReference(x) === a, c === 3),
+          testRelation),
+        Filter(And(OuterReference(y) === b, c === 6),
+          testRelation))
+    val correctAnswer =
+      Union(
+        Filter(And(x === a, c === 3),
+          DomainJoin(Seq(x, y),
+            testRelation)),
+        Filter(And(y === b, c === 6),
+          DomainJoin(Seq(x, y),
```

Comment:
So regardless of whether the condition is equality or non-equality, we need to insert domain joins here, right?

Reply:
Right, currently we always insert domain joins. We can optimize some equality cases in a follow-up, as I mentioned in the PR description